go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -854,19 +872,24 @@ func (npw *nodePortWatcher) DeleteEndpointSlice(epSlice *discovery.EndpointSlice
 
 	klog.V(5).Infof("Deleting endpointslice %s in namespace %s", epSlice.Name, epSlice.Namespace)
 	// remove rules for endpoints and add back normal ones
-	namespacedName, err := serviceNamespacedNameFromEndpointSlice(epSlice)
+	namespacedName, err := util.ServiceNamespacedNameFromEndpointSlice(epSlice)
 	if err != nil {
 		return fmt.Errorf("cannot delete %s/%s from nodePortWatcher: %v", epSlice.Namespace, epSlice.Name, err)
 	}
 	epSlices, err := npw.watchFactory.GetEndpointSlices(epSlice.Namespace, epSlice.Labels[discovery.LabelServiceName])
 	if err != nil {
-		return fmt.Errorf("could not fetch endpointslices for service %s during endpointSliceDelete: %v", namespacedName, err)
+		if !kerrors.IsNotFound(err) {
+			return fmt.Errorf("error retrieving all endpointslices for service %s/%s during endpointslice delete on %s: %w",
+				namespacedName.Namespace, namespacedName.Name, epSlice.Name, err)
+		}
+		klog.V(5).Infof("No endpointslices found for service %s/%s during endpointslice delete on %s: %v",
```

Comment:
Say there is one endpoint slice only for this service and it gets deleted. Since we're processing a delete event, we'll get a not found error when retrieving all endpoint slices for this service. It's the same logic that is used by level-driven controllers when they have to figure out whether to add/update or delete an object, for which all they know is its key: if they don't find it in the api server, they have to do a delete operation. 
What do you think?

If all that makes sense, I should actually set `epSlices = []*discovery.EndpointSlice{epSlice}` at the end of the if block, so that we actually remove this endpoint slice instead of returning early:
```
	epSlices, err := npw.watchFactory.GetEndpointSlices(epSlice.Namespace, epSlice.Labels[discovery.LabelServiceName])
	if err != nil {
		if !kerrors.IsNotFound(err) {
			return fmt.Errorf("error retrieving all endpointslices for service %s/%s during endpointslice delete on %s: %w",
				namespacedName.Namespace, namespacedName.Name, epSlice.Name, err)
		}
		klog.V(5).Infof("No other endpointslices found for service %s/%s during endpointslice delete on %s: %v",
			namespacedName.Namespace, namespacedName.Name, epSlice.Name, err)
		epSlices = []*discovery.EndpointSlice{epSlice}
	}
```